Cuda build

mfdeakin-sandia opened the pull request:
Use device views properly - don't attempt to access them from the host
Adds some fixes to compute_and_apply_rhs
Adds CMake build for Kokkos codes
Change constexpr values needed for at compile time by Kokkos to defines due to lack of CUDA support :-1:.
Redo the Region object to make accessing cleaner different variable views

mfdeakin-sandia committed:
Changes build to work with CMake

mfdeakin-sandia committed:
Support building with a prebuilt Kokkos, entirely through CMake.
Start of support building on a CUDA machine

mfdeakin-sandia committed:
Significant restructuring of globals to accomodate CUDA

mfdeakin-sandia committed:
Use scratch memory for more things. Also use the new constants interface

mfdeakin-sandia committed:
In progress transition of the sphere_operators to using scratch_memory instead of C arrays which can't be captured

mfdeakin-sandia committed:
Tons of changes to at least build on CUDA
  Can't use constexpr for compile time constants due to CUDA not supporting them correctly :(
  Temporarily disable printing and dumping results
  Only use device data, rather than share with the host
  Change Region structure and replace TestData structure
  Change structure name holding physical constants in anticipation of eventually using a structure for config constants (NP, NUM_LEVELS, ...)
  Make certain all functions callable from CUDA kernels are prefixed with KOKKOS_INLINE_FUNCTION or the like

Refactor of main.cpp to make it significantly cleaner/easier to read

mfdeakin-sandia committed:
Support linking with hwloc

mfdeakin-sandia committed:
Reduce the scope of several scratch views to save memory

mfdeakin-sandia committed:
Fix initialization of Control object views

mfdeakin-sandia committed:
Able to at least get to CUDA code now :)

mfdeakin-sandia committed:
Fix pressure initialization loop. Also rename p to pressure

mfdeakin-sandia committed:
Fix typos

mfdeakin-sandia committed:
Fix region.V calculation. Initialize T_vadv and v_vadv properly

bartgol commented:
I will take a look at this tomorrow.

bartgol commented:
I think the branch looks good. I want to fix the initialization, to get the same numbers of fortran and cxx/pointers_only, then we can merge.

bartgol committed:
Merge branch 'master' into cuda_build

bartgol committed:
Added a CMake option to specify whether kokkos was built via cmake or not.

If built with autotools, kokkos is condensed into a single library called
libkokkos, while when built with CMake, it generates libkokkoscore,
libkokkoscontainers, libkokkosalgorithms, libkokkostsqr.

bartgol committed:
Added a scratch memory manager class.

This class relies on some metaprogramming to do most of the work. Its main
job is to provide valid scratch memory pointers to the user so that he can
create scratch views as unmanaged view in the execution space. The user
needs to know in advance how many scratch views are needed; in particular,
the user must know how many views will be 'team' views (one per team) and
how many will be 'thread' views (one per thread, i.e., team member).

Then, the user must typedef a specialiation of CountAndSizePack for both
the team and thread sizes, and use these types to instantiate the scratch
memory manager.

The file ScratchMemoryDefs contains an example of how to create the manager,
by first defining (compile time) counters and sizes for each team/thread view
needed.

bartgol committed:
Removing a file that somehow ended up committed.

bartgol committed:
Using the ScratchManager to manage scratch memory in the kokkos_scratch build

bartgol committed:
Working on making fortran and cxx use and compute the same numbers

bartgol committed:
Fixed bug in vorticity_v2

The component u of the velocity was never used.

bartgol committed:
Added a macro in fortran to initialize phis state vector properly

bartgol committed:
Final fixes to the pointers_only version to make the numbers equal to those in fortran

bartgol committed:
Final fixes in kokkos_scratch to make the numbers equal to those in fortran

bartgol committed:
Making the number of elements a run time parameter in fortran

bartgol committed:
Some bug fixes to basic and kokkos_basic.

WARNING: basic still does not produce the same output, while kokkos_basic
         only produces the same output with 1 thread.
